File: LeetCode/Medium/0542-01-matrix/0542-01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        queue<pair<int,int>> q;
        int dy[4] = {1,-1,0,0};
        int dx[4] = {0,0,1,-1};
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(mat[i][j]==0)
                    q.push({i,j});
                else
                    mat[i][j] = m*n;
            }
        }
        while(!q.empty()){
            int y = q.front().first;
            int x = q.front().second;
            q.pop();
            for(int i=0; i<4; i++){
                int ny = y+dy[i];
                int nx = x+dx[i];
                if(0>ny || ny >= m)
                    continue;
                if(0>nx || nx >= n)
                    continue;
                if(mat[ny][nx]>mat[y][x]+1){
                    q.push({ny,nx});
                    mat[ny][nx] = mat[y][x]+1;
                }
            }
        }
        return mat;
    }
};

File: LeetCode/Medium/0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m = len(mat)
        n = len(mat[0])
        ans = [[0] * n for _ in range(m)]
        directions = [(0,1),(0,-1),(1,0),(-1,0)]
        q = collections.deque()

        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    q.append((i,j))
                else:
                    mat[i][j] = m*n
        
        while q:
            y, x = q.popleft()
            for dy, dx in directions:
                ny = y + dy
                nx = x + dx
                if 0<=ny<m and 0<=nx<n and mat[ny][nx] > mat[y][x] + 1:
                    mat[ny][nx] = mat[y][x] + 1
                    q.append((ny,nx))

        return mat